Update PyTorchStreamReader API to take cpu allocator override

huxintong opened the pull request:
Summary: Add allocator param in getRecord

Test Plan:
newly added UT
```
buck test caffe2/caffe2/serialize:inline_container_test
```

Differential Revision: D72252585

janeyx99 commented:
@albanD I'm surprised to find that we still have caffe2 features within our OSS repo.

albanD commented:
Part of the serialization c++ code is in the caffe2 folder. Which this is part of.

@huxintong we would need more context on why you're doing this before reviewing this. These APIs are not public so I don't expect anyone to use this as is?

huxintong commented:
@albanD The goal of this change is to use it in an internal tensor loader to support customized allocator at loading time for serving purpose, I put some context in the stacked diff. This change is not for public, using the open source tag just trying to find a reviewer

albanD commented on caffe2/serialize/inline_container.h:
```diff
@@ -130,11 +130,13 @@ class TORCH_API PyTorchStreamReader final {
   // return dataptr, size
-  std::tuple<at::DataPtr, size_t> getRecord(const std::string& name);
+  // set allocator to override default cpu allocator
+  std::tuple<at::DataPtr, size_t> getRecord(const std::string& name, at::Allocator* allocator = nullptr);
```
Please use std::optional<> for all functions

albanD reviewed (commented):
I'm not sure how well this is going to work going forward but why not.

albanD commented on caffe2/serialize/inline_container.cc:
```diff
@@ -371,7 +372,11 @@ std::tuple<at::DataPtr, size_t> PyTorchStreamReader::getRecord(
   mz_zip_reader_file_stat(ar_.get(), key, &stat);
   valid("retrieving file meta-data for ", name.c_str());
-  at::DataPtr retval = c10::GetCPUAllocator()->allocate(stat.m_uncomp_size);
+  at::Allocator* allocatorPtr = c10::GetCPUAllocator();
+  if (allocator.has_value()) {
+    allocatorPtr = allocator.value();
+  }
```
```suggestion
  at::Allocator* allocatorPtr = allocator.has_value() ? allocator.value() : c10::GetCPUAllocator();
```

To avoid the extra getcpuallocator call.
Same below

albanD reviewed (approved):
Please update the ternary, sounds good otherwise!

huxintong committed:
Update PyTorchStreamReader API to take cpu allocator override (#150439)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/150439

Add allocator param in getRecord

Test Plan:
newly added UT
```
buck test caffe2/caffe2/serialize:inline_container_test
```

Reviewed By: albanD

Differential Revision: D72252585
